Add Locker API

piscisaureus opened the pull request:
Fork of #738

piscisaureus commented:
@SpencerSharkey - I'm not trying to steal your thunder here, but I thought I'd try to move #738 forward a bit.

piscisaureus commented on src/snapshot.rs:
```diff
@@ -153,6 +163,30 @@ impl SnapshotCreator {
     }
   }
 
+  pub unsafe fn get_isolate_ptr(&self) -> *mut Isolate {
```
Does not need to be unsafe.

lbguilherme replied:
It looks like this method shouldn't be pub.

piscisaureus commented on src/snapshot.rs:
```diff
@@ -153,6 +163,30 @@ impl SnapshotCreator {
+
+  /// Borrows the Isolate from the SnapshotCreator, also returning a reference
+  /// to the SnapshotCreator.
+  pub fn get_isolate_and_handle(&mut self) -> (Locker, &mut SnapshotCreator) {
```
This method is a bit weird, I suppose it is needed to make `SnapshotCreator::set_default_context(context: Local<Context>)` work?
I think it would be better to change the signature to `SnapshotCreator::set_default_context(context: Global<Context>)`, that way it could be called outside a `HandleScope`.

lbguilherme replied:
I don't see the need for this method. The isolate can be obtained from `get_isolate` directly. And the snapshotcreator itself can be used for `set_default_context` just fine. The only requirement is that the `scope` must go out of scope (dropped) before `create_blob` is called. As far as I can tell, it can be removed.

lbguilherme commented on src/snapshot.rs:
```diff
@@ -153,6 +163,30 @@ impl SnapshotCreator {
+    function_code_handling: FunctionCodeHandling,
+  ) -> Option<StartupData> {
+    self.create_blob(function_code_handling)
+  }
```
Why was `into_blob` introduced instead of just using `create_blob`?

lbguilherme commented:
Bug: If an `IsolateHandle` is created with `v8::Isolate::new_handle`, the underlying isolate is never dropped because `drop_annex` is never called and the `Arc<IsolateAnnex>` has a cycle. Additionally, there is no way to drop the annex using only the public API.

merlinfuchs commented:
Really looking forward for this to be done. 👍  Sadly not experienced enough to help with this myself ...

merlinfuchs commented on src/isolate.rs:
```diff
@@ -729,79 +751,140 @@ impl Isolate {
+
+  /// Borrows a new [`Locker`] reference for the Isolate. While Locked, the
+  /// embedder is free to interact with an Isolate on the current thread.
+  pub fn lock(&mut self) -> Locker {
```
Why does this require a mutual reference? It does not mutate anything and also doesn't require exclusive access because `IsolateHandle` can be cloned, right?

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/denoland/rusty_v8?pullRequest=788) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/denoland/rusty_v8?pullRequest=788) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/denoland/rusty_v8?pullRequest=788) it.</sub>

piscisaureus commented:
I just rebased this PR and fixed the tests again.

We are not merging this PR at this time because is has unaddressed safety issues and at least one memory leak.
In Deno and Deno Deploy we don't need to access isolates from different threads so it's hard to evaluate whether this PR addresses the need.

We would appreciate help from the community to get this PR into a shape where we can land it.

ry commented:
closing because stale.
